packages/backend/src/open_payments/auth/middleware.ts

```diff
@@ -0,0 +1,134 @@
+import assert from 'assert'
+import axios from 'axios'
+
+import {
+  AccessType,
+  AccessAction,
+  Grant,
+  GrantOptions,
+  GrantAccess,
+  GrantAccessJSON
+} from './grant'
+import { AppContext } from '../../app'
+
+export function createAuthMiddleware({
+  type,
+  action
+}: {
+  type: AccessType
+  action: AccessAction
+}) {
+  return async (
+    ctx: AppContext,
+    next: () => Promise<unknown>
+  ): Promise<void> => {
+    try {
+      if (
+        !ctx.request.headers.authorization ||
+        ctx.request.headers.authorization.split(' ')[0] !== 'GNAP'
+      ) {
+        ctx.throw(401, 'Unauthorized')
+      }
+      const token = ctx.request.headers.authorization.split(' ')[1]
+      const config = await ctx.container.use('config')
+      const grant = await getGrant(config.authServerIntrospectionUrl, token)
+      if (!grant || !grant.active) {
+        ctx.throw(401, 'Invalid Token')
+      }
+      if (
+        !grant.includesAccess({
+          type,
+          action,
+          identifier: ctx.params.accountId
+        })
+      ) {
+        ctx.throw(403, 'Insufficient Grant')
+      }
+      ctx.grant = grant
+      await next()
+    } catch (err) {
+      if (err.status === 401) {
+        ctx.status = 401
+        ctx.message = err.message
+        const config = await ctx.container.use('config')
```

Comment:
nit: This can go above the `try { ...` block, so that the other `config` declaration (line 33) can be removed.